### PR TITLE
Repository Async Implementation For Create Methods

### PR DESCRIPTION
**Pull request by klauffer**
implemented all of the Create methods as async

### CODE REVIEW

**Review by jstevenson — APPROVED** [review]

**Comment by jstevenson** [comment]
Should `public virtual IResult<IEnumerable<T>> BulkCreateOrUpdate()` and all other methods referencing other obsolete methods, also be obsolete?

File: src/Core/Interfaces/Data/IRepositoryAsync.cs [hunk 1]
```diff
@@ -7,13 +9,122 @@ namespace AndcultureCode.CSharp.Core.Interfaces.Data
...
+        /// <param name="property">Property or properties of the typed object to determine distinctness</param>
+        /// <param name="createdById">Id of the user creating the entity</param>
+        /// <param name="cancellationToken">a token allowing aborting of this request</param>
+        /// <typeparam name="TKey"></typeparam>
+        /// <returns></returns>
+        Task<IResult<List<T>>> CreateDistinctAsync<TKey>(IEnumerable<T> items, Func<T, TKey> property, long? createdById = null, CancellationToken cancellationToken = default);
+
+        #endregion Create
+
+        #region Delete
```
Comment (jstevenson) [inline src/Core/Interfaces/Data/IRepositoryAsync.cs:67]: Minor: space after

File: test/Conductors.Tests/RepositoryConductorTests/DeleteAsyncShould.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        [Fact]
+        public async Task Succeed_When_Given_Proper_Id()
+        {
+            // Arrange 
+            var repositoryMock = new RepositoryMock<UserStub>()
+                .DeleteAsync(new Result<bool>(true));
+            var respositoryConductor = SetupSut(repositoryMock);
+
+
```
Comment (jstevenson) [inline test/Conductors.Tests/RepositoryConductorTests/DeleteAsyncShould.cs:40]: Minor: The spacing before and after Arrange/Act/Assert comments is a tad inconsistent in this file. My pattern-matching brain is breaking ;)

**Review by jstevenson — APPROVED** [review]
Changes look great!

File: test/Conductors.Tests/RepositoryConductorTests/BulkCreateAsyncShould.cs [hunk 3]
```diff
@@ -52,12 +44,45 @@ public async Task Throw_Argument_Null_Exception_When_Given_Null_Input()
...
+        [Fact]
+        public async Task Throw_OperationCanceledException_If_Canceled()
+        {
+            // Arrange 
+            var repositoryMock = new RepositoryMock<UserStub>();
+            var respositoryConductor = SetupSut(repositoryMock);
+            var cancellationTokenSource = new CancellationTokenSource();
+            var cancellationToken = cancellationTokenSource.Token;
+            cancellationTokenSource.Cancel();
+            // Act & Assert
```
Comment (myty) [inline test/Conductors.Tests/RepositoryConductorTests/BulkCreateAsyncShould.cs:63]: nitpick: I won't call it out everywhere, but consider adding line spaces above code comments like this for readability.

File: test/Conductors.Tests/RepositoryConductorTests/BulkDeleteAsyncShould.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                .BulkDeleteAsync(new Result<bool>(true));
+            var respositoryConductor = SetupSut(repositoryMock);
+            var userStubs = new List<UserStub>()
+            {
+                new UserStub()
+            };
+
+            // Act
+            var result = await respositoryConductor.BulkDeleteAsync(userStubs);
+            // Assert
```
Comment (myty) [inline test/Conductors.Tests/RepositoryConductorTests/BulkDeleteAsyncShould.cs:46]: nitpick: line spacing

**Review by myty — COMMENTED** [review]
Left a comment about passing a `cancellationToken`, but other than that, it's just cosmetic nitpicky stuff.  👍🏻

File: test/Conductors.Tests/RepositoryConductorTests/BulkDeleteAsyncShould.cs (hunk 4 above)
Comment (klauffer) [reply test/Conductors.Tests/RepositoryConductorTests/BulkDeleteAsyncShould.cs:47]: spacing resolved

File: test/Conductors.Tests/RepositoryConductorTests/BulkCreateAsyncShould.cs (hunk 3 above)
Comment (klauffer) [reply test/Conductors.Tests/RepositoryConductorTests/BulkCreateAsyncShould.cs:64]: spacing resolved

File: src/Conductors/RepositoryCreateConductorAsync.cs [hunk 5]
```diff
@@ -36,11 +37,47 @@ IRepository<T> repository
...
+            if (!items.Any()) throw new ArgumentException(ARGUMENT_EXCEPTION_MESSAGE, nameof(items));
+            return _repository.BulkCreateDistinctAsync(items, property, createdById, cancellationToken);
+        }
+
+        /// <inheritdoc />
+        public virtual Task<IResult<T>> CreateAsync(T item, long? createdById = null, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (item == null) throw new ArgumentNullException(nameof(item));
+            return _repository.CreateAsync(item, createdById);
```
[the remark this answers is not in the document]
Comment (klauffer) [reply src/Conductors/RepositoryCreateConductorAsync.cs]: ...yep, totally missed that

File: src/Conductors/RepositoryCreateConductorAsync.cs [hunk 6]
```diff
@@ -36,11 +37,47 @@ IRepository<T> repository
...
         /// <inheritdoc />
-        public virtual Task<IResult<List<T>>> BulkCreateAsync(IEnumerable<T> items, long? createdById = null) 
+        public virtual Task<IResult<List<T>>> BulkCreateAsync(IEnumerable<T> items, long? createdById = null, CancellationToken cancellationToken = default) 
         {
-            if(items == null) throw new ArgumentNullException(nameof(items));
-            if(!items.Any()) throw new ArgumentException("An empty collection was provided", nameof(items));
+            cancellationToken.ThrowIfCancellationRequested();
+            if (items == null) throw new ArgumentNullException(nameof(items));
+            if(!items.Any()) throw new ArgumentException(ARGUMENT_EXCEPTION_MESSAGE, nameof(items));
             return _repository.BulkCreateAsync(items, createdById);
```
[the remark this answers is not in the document]
Comment (klauffer) [reply src/Conductors/RepositoryCreateConductorAsync.cs]: ...yep, totally missed that

File: test/Conductors.Tests/RepositoryConductorTests/DeleteAsyncShould.cs (hunk 2 above)
Comment (klauffer) [reply test/Conductors.Tests/RepositoryConductorTests/DeleteAsyncShould.cs:40]: your pattern matching brain should now be soothed :D

File: src/Core/Interfaces/Data/IRepositoryAsync.cs (hunk 1 above)
Comment (klauffer) [reply src/Core/Interfaces/Data/IRepositoryAsync.cs:67]: space invaded

File: src/Conductors/RepositoryConductorAsync.cs [hunk 7]
```diff
@@ -72,6 +72,7 @@ public partial class RepositoryConductor<T> : Conductor, IRepositoryConductor<T>
 
         #endregion Constructor
 
+        #region Create
```
[the remark this answers is not in the document]
Comment (klauffer) [reply src/Conductors/RepositoryConductorAsync.cs:75]: urgent matter resolved as quickly as bureaucracy

**Review by myty — APPROVED** [review]
looks good!

**Pull request merged**